src/utils/chain_ops.js

```diff
@@ -0,0 +1,137 @@
+}
+
+// Fetch a block by block number or block hash
+export async function getBlock(api, numberOrHash) {
+  const hash = isHexWithGivenByteSize(numberOrHash, 32) ? numberOrHash : (await blockNumberToHash(api, numberOrHash));
+  const { block } = await api.rpc.chain.getBlock(hash);
+  return block;
+}
+
+// Given a block header or block, extract the block number
+export function getBlockNo(headerOrBlock) {
+  if ((typeof headerOrBlock.header) === 'object') {
+    return headerOrBlock.header.number.toNumber();
+  }
+  return headerOrBlock.number.toNumber();
+}
+
+// Given a block number or block hash, get all extrinsics.
+export async function getAllExtrinsicsFromBlock(api, numberOrHash) {
+  const block = await getBlock(api, numberOrHash);
+  return block.extrinsics;
+}
+
+// Given a block number or block hash, get all events.
+export async function getAllEventsFromBlock(api, numberOrHash, formatted = true) {
+  const hash = isHexWithGivenByteSize(numberOrHash, 32) ? numberOrHash : (await blockNumberToHash(api, numberOrHash));
+  const events = await api.query.system.events.at(hash);
+  return formatted ? events.map((e) => e.toJSON()) : events;
+}
+
+// Format a balance with units or only as number.
+function formatBalIfNeeded(bal, format = true) {
+  return format ? formatBalance(bal, { decimals: DECIMAL, withSi: true, withUnit: SYMBOL }) : bal.toNumber();
+}
+
+// Given a block number or block hash, get all transfer extrinsics. Each returned extrinsic is in form `[sender, recipient, amount]`
+export async function getTransfersFromBlock(api, numberOrHash, network, balanceFormatted = true) {
+  const extrinsics = await getAllExtrinsicsFromBlock(api, numberOrHash);
+  const transfers = [];
+  extrinsics.forEach((ext) => {
+    if (ext.method && ext.method.section === 'balances' && (ext.method.method === 'transfer' || ext.method.method === 'transferKeepAlive')) {
+      const bal = formatBalIfNeeded(ext.method.args[1], balanceFormatted);
+      transfers.push([asDockAddress(ext.signer, network), asDockAddress(ext.method.args[0], network), bal]);
+    }
+  });
+  return transfers;
+}
+
+// Get free balance of an account.
+export async function getBalance(api, address, formatted = true) {
+  const { data: { free } } = await api.query.system.account(address);
+  return formatBalIfNeeded(free, formatted);
+}
+
+function checkValidMicroAmount(micros) {
+  // TODO: this should handle upto 64 bits
+  if (!Number.isSafeInteger(micros) || micros < 0) {
+    throw new Error(`Amount must be a positive integer and must fit in 53 bits but was ${micros}`);
+  }
+}
+
+// Convert tokens to micros
+export function docksToMicroDocks(amount) {
+  // Token has `DECIMAL` decimal places
+  const micros = amount * (10 ** DECIMAL);
+  checkValidMicroAmount(micros);
+  return micros;
+}
+
+// Transfer tokens and return txn hash. Converts to mirco tokens first
+export async function transferDocks(api, senderKeypair, recipAddr, amount) {
```

Comment:
should be `transferDock` not `transferDocks` ? same with below for `MicroDocks`. We would say, for example: "Can you transfer me 1000 DOCK" not "DOCKs"